Source/Tests/LinuxSyscalls/SyscallsVMATracking.cpp

```diff
@@ -0,0 +1,367 @@
+    LOGMAN_THROW_A_FMT(VMA->ResourceNextVMA->ResourcePrevVMA == VMA, "VMA tracking error");
+    VMA->ResourceNextVMA->ResourcePrevVMA = NewVMA;
+  }
+
+  ListCheckVMALinks(VMA);
+  ListCheckVMALinks(NewVMA);
+  ListCheckVMALinks(VMA->ResourceNextVMA);
+  ListCheckVMALinks(VMA->ResourcePrevVMA);
+}
+
+// Inserts a VMA in corresponding MappedResource list
+// Requires NewVMA->Resource, NewVMA->ResourcePrevVMA and NewVMA->ResourceNextVMA to be already setup
+void SyscallHandler::VMATracking::ListInsert(VMAEntry *AfterVMA, VMAEntry *NewVMA) {
+  LOGMAN_THROW_A_FMT(NewVMA->Resource != nullptr, "VMA tracking error");
+
+  LOGMAN_THROW_A_FMT(AfterVMA->Resource == NewVMA->Resource, "VMA tracking error");
+  LOGMAN_THROW_A_FMT(NewVMA->ResourcePrevVMA == AfterVMA, "VMA tracking error");
+  LOGMAN_THROW_A_FMT(NewVMA->ResourceNextVMA == AfterVMA->ResourceNextVMA, "VMA tracking error");
+
+  if (AfterVMA->ResourceNextVMA) {
+    LOGMAN_THROW_A_FMT(AfterVMA->ResourceNextVMA->ResourcePrevVMA == AfterVMA, "VMA tracking error");
+    AfterVMA->ResourceNextVMA->ResourcePrevVMA = NewVMA;
+  }
+  AfterVMA->ResourceNextVMA = NewVMA;
+
+  ListCheckVMALinks(AfterVMA);
+  ListCheckVMALinks(NewVMA);
+  ListCheckVMALinks(AfterVMA->ResourceNextVMA);
+  ListCheckVMALinks(AfterVMA->ResourcePrevVMA);
+}
+
+// Prepends a VMA
+// Requires NewVMA->Resource, NewVMA->ResourcePrevVMA and NewVMA->ResourceNextVMA to be already setup
+void SyscallHandler::VMATracking::ListPrepend(MappedResource *Resource, VMAEntry *NewVMA) {
+  LOGMAN_THROW_A_FMT(Resource != nullptr, "VMA tracking error");
+
+  LOGMAN_THROW_A_FMT(NewVMA->Resource == Resource, "VMA tracking error");
+  LOGMAN_THROW_A_FMT(NewVMA->ResourcePrevVMA == nullptr, "VMA tracking error");
+  LOGMAN_THROW_A_FMT(NewVMA->ResourceNextVMA == Resource->FirstVMA, "VMA tracking error");
+
+  if (Resource->FirstVMA) {
+    LOGMAN_THROW_A_FMT(Resource->FirstVMA->ResourcePrevVMA == nullptr, "VMA tracking error");
+    Resource->FirstVMA->ResourcePrevVMA = NewVMA;
+  }
+
+  Resource->FirstVMA = NewVMA;
+
+  ListCheckVMALinks(NewVMA);
+  ListCheckVMALinks(NewVMA->ResourceNextVMA);
+  ListCheckVMALinks(NewVMA->ResourcePrevVMA);
+}
+
+/// VMA tracking ///
+
+// Lookup a VMA by address
+SyscallHandler::VMATracking::VMAsType::const_iterator SyscallHandler::VMATracking::LookupVMAUnsafe(uint64_t GuestAddr) const {
+  auto Entry = VMAs.upper_bound(GuestAddr);
+
+  if (Entry != VMAs.begin()) {
+    --Entry;
+
+    if (Entry->first <= GuestAddr && (Entry->first + Entry->second.Length) > GuestAddr) {
+      return Entry;
+    }
+  }
+
+  return VMAs.end();
+}
+
+// Set or Replace mappings in a range with a new mapping
+void SyscallHandler::VMATracking::SetUnsafe(FEXCore::Context::Context *CTX, MappedResource *MappedResource, uintptr_t Base,
+                                            uintptr_t Offset, uintptr_t Length, VMAOptions Flags) {
+  ClearUnsafe(CTX, Base, Length, MappedResource);
+
+  auto VMAInserted = VMAs.emplace(
```

Comment:
Oh yeah, we're using new enough C++! I've used this features a few times and then had to back-down because of compiler issues in the past, so kind of was in the "nice to have but too new" bucket.